RecoParticleFlow/PFClusterProducerAlpaka/plugins/alpaka/PFClusterProducerKernel.dev.cc

```diff
@@ -0,0 +1,2224 @@
+                          pfClusParams.endcapTimeResConsts_corrTermLowE() / (energy * energy);
+        res2 = res * res;
+      } else {
+        const float noiseDivE = pfClusParams.endcapTimeResConsts_noiseTermLowE() / energy;
+        res2 = noiseDivE * noiseDivE + pfClusParams.endcapTimeResConsts_constantTermLowE2();
+      }
+    } else if (energy < pfClusParams.endcapTimeResConsts_threshHighE()) {
+      const float noiseDivE = pfClusParams.endcapTimeResConsts_noiseTerm() / energy;
+      res2 = noiseDivE * noiseDivE + pfClusParams.endcapTimeResConsts_constantTerm2();
+    } else  // if (energy >=threshHighE_)
+      res2 = pfClusParams.endcapTimeResConsts_resHighE2();
+
+    if (res2 > 10000.)
+      return 10000.;
+    return res2;
+  }
+
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE float timeResolution2Barrel(
+      reco::PFClusterParamsAlpakaESDataDevice::ConstView pfClusParams, const float energy) {
+    float res2 = 10000.;
+
+    if (energy <= 0.)
+      return res2;
+    else if (energy < pfClusParams.barrelTimeResConsts_threshLowE()) {
+      if (pfClusParams.barrelTimeResConsts_corrTermLowE() > 0.) {  // different parametrisation
+        const float res = pfClusParams.barrelTimeResConsts_noiseTermLowE() / energy +
+                          pfClusParams.barrelTimeResConsts_corrTermLowE() / (energy * energy);
+        res2 = res * res;
+      } else {
+        const float noiseDivE = pfClusParams.barrelTimeResConsts_noiseTermLowE() / energy;
+        res2 = noiseDivE * noiseDivE + pfClusParams.barrelTimeResConsts_constantTermLowE2();
+      }
+    } else if (energy < pfClusParams.barrelTimeResConsts_threshHighE()) {
+      const float noiseDivE = pfClusParams.barrelTimeResConsts_noiseTerm() / energy;
+      res2 = noiseDivE * noiseDivE + pfClusParams.barrelTimeResConsts_constantTerm2();
+    } else  // if (energy >=threshHighE_)
+      res2 = pfClusParams.barrelTimeResConsts_resHighE2();
+
+    if (res2 > 10000.)
+      return 10000.;
+    return res2;
+  }
+
+  // Calculation of dR2 for Clustering
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE float dR2(Position4 pos1, Position4 pos2) {
+    float mag1 = sqrtf(pos1.x * pos1.x + pos1.y * pos1.y + pos1.z * pos1.z);
+    float cosTheta1 = mag1 > 0.0 ? pos1.z / mag1 : 1.0;
+    float eta1 = 0.5 * logf((1.0 + cosTheta1) / (1.0 - cosTheta1));
+    float phi1 = atan2f(pos1.y, pos1.x);
+
+    float mag2 = sqrtf(pos2.x * pos2.x + pos2.y * pos2.y + pos2.z * pos2.z);
+    float cosTheta2 = mag2 > 0.0 ? pos2.z / mag2 : 1.0;
+    float eta2 = 0.5 * logf((1.0 + cosTheta2) / (1.0 - cosTheta2));
+    float phi2 = atan2f(pos2.y, pos2.x);
+
+    float deta = eta2 - eta1;
+    float dphi = fabsf(fabsf(phi2 - phi1) - PI_F) - PI_F;
+    return (deta * deta + dphi * dphi);
+  }
+
+  // Helper function used in atomicMaxF
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE int floatAsInt(float fval) {
+    union {
+      float fval;
+      int ival;
+    } u;
+
+    u.fval = fval;
+
+    return u.ival;
```

Comment:
CUDA and HIP/ROCm provide `__device__` functions for this: `float __int_as_float(int x)` and `int __float_as_int(float x)`.

Reply:
By the way, we discussed this with the Alpaka developers in https://github.com/alpaka-group/alpaka/issues/1959, the idea would be to eventually implement the missing atomic operations directly in alpaka.

Reply:
Is there a way to still access and use the CUDA and HIP/ROCM functions in Alpaka for this then? And I guess one solution if so is to also have a condition for CPU to instead use the `edm::bit_cast`?

Reply:
The recommended way is to write a function object that takes `TAcc const& acc` as its first argument, and then specialise it for the different accelerator types.

A simpler but less extensible approach is to use various `#ifdef` to implement different versions for CUDA, HIP/ROCm, and CPU.

Reply:
I am unsure of how to use CUDA/HIP/ROCm `__device__` functions here since `ALPAKA_FN_ACC` sets the function to `__host__ __device__` without specifying `ALPAKA_ACC_GPU_CUDA_ONLY_MODE`. My solution now is:
```C++
template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
ALPAKA_FN_ACC ALPAKA_FN_INLINE static float atomicMaxF(const TAcc& acc, float* address, float val) {
    if constexpr (not requires_single_thread_per_block_v<TAcc>) {
        // GPU implementation uses hacky intAsFloat and floatAsInt functions
        int ret = floatAsInt(*address);
        while (val > intAsFloat(ret)) {
            int old = ret;
            if ((ret = alpaka::atomicCas(acc, (int*)address, old, floatAsInt(val))) == old)
                break;
        }
        return intAsFloat(ret);
    } else {
        // CPU implementation uses edm::bit_cast
        int ret = edm::bit_cast<int>(*address);
        while (val > edm::bit_cast<float>(ret)) {
            int old = ret;
            if ((ret = alpaka::atomicCas(acc, (int*)address, old, edm::bit_cast<int>(val))) == old)
                break;
        }
        return edm::bit_cast<float>(ret);
    }
}
```

I am just missing something to get the official `__float_as_int` and `__int_as_float` functions to work?

Reply:
I would not use `requires_single_thread_per_block_v<TAcc>` here, because we can foresee CPU-based accelerators for which the trait is `false`, and GPU-based accelerators that do not support those functions.

Since these are really back-end specific, I think it would be reasonable to use:
```c++
#if defined(__CUDA_ARCH__) or defined (__HIP_DEVICE_COMPILE__)
  // CUDA/HIP version
#else
  // CPU version
#endif
```

Reply:
Thanks @fwyzard, I'll make the change.